python/ray/serve/api.py

```diff
@@ -1235,3 +1169,179 @@ def list_deployments() -> Dict[str, Deployment]:
         )
 
     return deployments
+
+
+def deploy_group(
+    deployments: List[Deployment],
+    _blocking: bool = True,
+    controller: ServeController = None,
+) -> List[GoalId]:
+    """
+    EXPERIMENTAL API
+
+    Takes in a list of deployment object, and deploys them atomically.
+
+    Args:
+        deployments(List[Deployment]): a list of deployments to deploy.
+        _blocking(bool): whether to wait for the deployments to finish
+            deploying or not.
+        controller: the Serve controller to use when deploying the deployments.
+            If None, uses the global_client's controller.
+    """
+
+    deployment_args_list = []
+    for deployment in deployments:
+        deployment_args_list.append(
+            _get_deploy_args(
+                deployment._name,
+                deployment._func_or_class,
+                deployment.init_args,
+                deployment.init_kwargs,
+                ray_actor_options=deployment._ray_actor_options,
+                config=deployment._config,
+                version=deployment._version,
+                prev_version=deployment._prev_version,
+                route_prefix=deployment.route_prefix,
+            )
+        )
+
+    if controller is None:
+        controller = _get_global_client()._controller
+    update_goals = ray.get(controller.deploy_group.remote(deployment_args_list))
```

Comment:
why not implementing `deploy_group` in `Client` so that you can reuse its instance variables ?

Reply:
I moved all of `deploy_group`'s helper functions into the `Client` class.

However, `deploy_group` still has some heavy-weight logic left outside the `Client` class. What are your thoughts on making another `deploy_group` function inside the `Client` class and using the global `deploy_group` function as a thin wrapper that parses the `Deployments` and calls the `Client`'s `deploy_group`? This would be similar to how the global `deploy` function and the `Client`'s `deploy` function work together.

Note that one possible downside is that the `Client` class can't directly interface with `Deployment` objects. That means the `Client`'s `deploy_group` would need to take in a list of dictionaries containing the properties for every `Deployment` instead of a list of `Deployment`s. The UX for the end user would still be the same (passing in a list of `Deployment`s) since they'll interface with the global `deploy_group`. However, the internal code may look a bit messier.

Reply:
Another `deploy_group` inside `Client` w/ a global thin wrapper sounds good to me as it matches the others.